### DISCUSSION TITLE
Trying to run a series of bash commands on Windows but failed

### DISCUSSION THREAD

**Discussion by citystrawman** [Tyrrrz/CliWrap#236] [Q&A] [answered]
Hi, I am trying to run a series of bash commands on Windows but failed. 
the bash commands contains 3 steps, and I checked its validity by typing the commands manually:
1. I started bash by typing bash into powershell;
2. after entering bash, I ran the 3 commands in one round by concatenating the 3 commands using `;`, the result is as follows:
![image](https://github.com/Tyrrrz/CliWrap/assets/9004748/d8c10288-db43-4d18-b617-7a570c75718d)

In this sense, I think CliWrap could run these commands, and I implemented is as follows:
```
var result = await Cli.Wrap(bashPath)
    .WithArguments(args => args
    .Add($"export FREESURFER_HOME={sourceScriptDirWsl}")
    .Add(" ; ")
    .Add("source $FREESURFER_HOME/SetUpFreeSurfer.sh")
    .Add(" ; ")
    .Add(stripPathWsl)
    .Add("-i")
    .Add(inputPathWsl)
    .Add("-o")
    .Add(outputPathWsl)
    ).ExecuteBufferedAsync(cts.Token);
```
However, when running this code, I got errors:

```
Command execution failed, see the inner exception for details.

Standard error:
/bin/bash: export FREESURFER_HOME=/mnt/e/MRI-WPF/AlphaBrain/bin/Debug/net8.0-windows/Commands/freesurfer: No such file or directory
```

and the command is as follows::

`Command = {C:\windows\system32\bash.exe "export FREESURFER_HOME=/mnt/e/freesurfer" " ; " "source $FREESURFER_HOME/SetUpFreeSurfer.sh" " ; " /mnt/e/freesurfer/bin/mri_synthstrip -i /mnt/e/T1_t1_mprage.nii -o /mnt/e/stripped.nii}`

I am not sure what is the problem here. I am not sure if CliWrap could not handle **multiple sequenced** command, or whether the environment variable FREESURFER_HOME could not be regonized in string format?
I appreciate if you could help me solving this issue. Thank you!

**Comment by citystrawman** [comment]
Now, I even tried hard coding:

```
string argstr = "export FREESURFER_HOME=/mnt/e/freesurfer ; source $FREESURFER_HOME/SetUpFreeSurfer.sh ; /mnt/e/freesurfer/bin/mri_synthstrip -i /mnt/e/T1_t1_mprage.nii -o /mnt/e/strippedBash.nii";

var result = await Cli.Wrap(bashPath)
    .WithArguments(args => args
    .Add(argstr)
    ).ExecuteBufferedAsync(cts.Token);
```
but still got same errors. However, if I paste the argstr manually into bash, it runs successfully.

**Comment by Tyrrrz** [comment]
I'm pretty sure you need to do `bash -c "<your command>"`, so something like this:

```csharp
var result = await Cli.Wrap(bashPath)
    .WithArguments(args => args
        .Add("-c")
        .Add(new ArgumentsBuilder()
		    .Add($"export FREESURFER_HOME={sourceScriptDirWsl}")
		    .Add(" ; ")
		    .Add("source $FREESURFER_HOME/SetUpFreeSurfer.sh")
		    .Add(" ; ")
		    .Add(stripPathWsl)
		    .Add("-i")
		    .Add(inputPathWsl)
		    .Add("-o")
		    .Add(outputPathWsl)
            .Build()
       )
    ).ExecuteBufferedAsync(cts.Token);
```

**Reply by citystrawman** [reply]
Hi, I tried this method, but CliWrap still reports `No such file or directory` when it executes at $"export FREESURFER_HOME={sourceScriptDirWsl}" 

![image](https://github.com/Tyrrrz/CliWrap/assets/9004748/37cd9463-d4fc-4a5a-9350-e5bfdc18b352)

![image](https://github.com/Tyrrrz/CliWrap/assets/9004748/3cfa13d5-be21-4c18-b48b-0c4508037983)

the above screenshots are error info and the directory that I need to export as FREESURFER_HOME 

![image](https://github.com/Tyrrrz/CliWrap/assets/9004748/e95fcb7e-e63b-4dfc-9fb8-e59fcf064afa)
and here's my CliWrap code.

**Reply by Tyrrrz** [reply]
Can you try this instead:

```csharp
var result = await Cli.Wrap(bashPath)
    .WithArguments(args => args
        .Add("-c")
        .Add(new ArgumentsBuilder()
		    .Add("source $FREESURFER_HOME/SetUpFreeSurfer.sh")
		    .Add(" ; ")
		    .Add(stripPathWsl)
		    .Add("-i")
		    .Add(inputPathWsl)
		    .Add("-o")
		    .Add(outputPathWsl)
            .Build()
        )
    ).WithEnvironmentVariables(env => env
        .Set("FREESURFER_HOME", sourceScriptDirWsl)
    )
).ExecuteBufferedAsync(cts.Token);
```

**Reply by citystrawman** [reply, marked as answer]
Thank you! I found that by using `new ArgumentsBuilder()` the string has double quotes, thus I just use hard-coded string and now it works. 
![image](https://github.com/Tyrrrz/CliWrap/assets/9004748/3577d940-da37-4369-8f18-a066c28ceb3b)

Thank you so much!
